File: core/03 - OOB Tutorials/03 - OOB Messaging/OobMessaging_ViewController.m
/*
 * Copyright © 2020-2023 THALES. All rights reserved.
 */

// IMPORTANT: This source code is intended to serve training information purposes only.
//            Please make sure to review our IdCloud documentation, including security guidelines.

#import "OobMessaging_ViewController.h"
#import "OobMessaging_Logic.h"
#import "Provisioning_Logic.h"

@interface OobMessaging_ViewController()

@property (weak, nonatomic) IBOutlet UIButton *buttonFetchMessages;

@end

// Review: SNE document private methods
@implementation OobMessaging_ViewController

// MARK: - Life Cycle

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    
    [OobMessaging_Logic setCurrenOobListener:self];
}

- (void)viewWillDisappear:(BOOL)animated
{
    [super viewWillDisappear:animated];
    
    [OobMessaging_Logic setCurrenOobListener:nil];
}

// MARK: - Override

- (id<EMOathToken>)updateGUI
{
    // All token related error is solved in previous chapter.
    id<EMOathToken> token = [super updateGUI];
    
    // To make demo simple we will just disable / enable UI.
    [_buttonFetchMessages setEnabled:![self loadingBarIsPresent] && token];
    
    return token;
}


// MARK: - OobMessageListener

- (void)onOobLoadingShow:(NSString *)caption
{
    [self loadingBarShow:caption animated:YES];
}

- (void)onOobLoadingHide
{
    [self loadingBarHide:YES];
}

- (void)onOobDisplayMessage:(NSString *)message
{
    [self displayMessageDialog:message];
}

- (void)onOobApproveMessage:(NSString *)message
        withServerChallenge:(id<EMSecureString>)serverChallenge
          completionHandler:(void (^)(OTP_Value *otp))handler
{
    // Prepare simple dialog with two actions.
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:NSLocalizedString(@"STRING_PUSH_APPROVE_QUESTION", nil)
                                                                   message:message
                                                            preferredStyle:UIAlertControllerStyleAlert];
    
    // First is for approving. In that case we need to calculate OTP.
    [alert addAction:[UIAlertAction actionWithTitle:NSLocalizedString(@"STRING_PUSH_APPROVE_QUESTION_APPROVE", nil)
                                              style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
                                                  [self approveMessage:serverChallenge completionHandler:handler];
                                              }]];
    
    // Second is deny action. For such message we don't need to calculate OTP.
    [alert addAction:[UIAlertAction actionWithTitle:NSLocalizedString(@"STRING_PUSH_APPROVE_QUESTION_DENY", nil)
                                              style:UIAlertActionStyleDestructive handler:^(UIAlertAction * _Nonnull action) {
                                                  handler(nil);
                                              }]];
    
    // Display messagebox and lock thread to wait for response.
    [self presentViewController:alert animated:YES completion:nil];
}

// MARK: - User Interface

- (IBAction)onButtonPressedFetchMessages:(UIButton *)sender
{
    // Any incoming messages are handled through OobMessageListener as normal push notification.
    [OobMessaging_Logic fetchMessages];
}

// MARK: - Private Helpers

- (void)approveMessage:(id<EMSecureString>)serverChallenge
     completionHandler:(void (^)(OTP_Value *otp))handler
{
    // Get pin from user.
    [self userPinWithCompletionHandler:^(id<EMPinAuthInput> pin) {
        NSError     *internalErr    = nil;
        OTP_Value   *otpValue       = nil;
        
        // Server challenge is used for ocra calculation during transaction sign.
        // Without challenge it's authentication request.
        if (serverChallenge) {
            otpValue = [OobMessaging_Logic generateOtp:[Provisioning_Logic token]
                                         withAuthInput:pin
                                       serverChallenge:serverChallenge
                                                 error:&internalErr];
        } else {
            otpValue = [OTP_Logic generateOtp:[Provisioning_Logic token]
                                withAuthInput:pin
                                        error:&internalErr];
        }
        
        
        // We can return control to handler.
        // If there is some error present and OTPValue is nil, we still want to send response to server. In this case reject.
        handler(otpValue);
        
        // All secure containers should be wiped asap.
        [pin wipe];
        [otpValue wipe];
        
        // Display error if something went wrong.
        if (internalErr) {
            [self displayMessageDialogError:internalErr];
        }
    }];
}


@end

File: fasttrack/03 - OOB Tutorials/03 - OOB Messaging/OOBMessagingViewController.swift
/*
 * Copyright © 2020-2023 THALES. All rights reserved.
 */

// IMPORTANT: This source code is intended to serve training information purposes only.
//            Please make sure to review our IdCloud documentation, including security guidelines.

class OOBMessagingViewController: OOBRegistrationViewController {
    @IBOutlet weak var fetchButton:UIButton!
    var oOBMessagingLogic = OOBMessagingLogic()
    
    override func updateGui() {
        super.updateGui()
        let token = ProvisioningLogic.getToken()
        fetchButton.isEnabled = (token != nil)
    }
    
    @IBAction func fetchButtonPressed(_ sender:Any) {
        oOBMessagingLogic.fetchMessage { (response, error) in
            if let err = error {
                self.displayMessageDialogError(error: err)
            }
            if let oobResponse = response {
                if oobResponse.hasIncomingMessage {
                    if oobResponse.messageType == .unsupported {
                        self.displayMessageDialog(result: "Fetch message is successful but message type not supported")
                    } else {
                        self.handleOOBIncomingMessage(oobResponse)
                    }
                }  else {
                    self.displayMessageDialog(result: "No messages")
                }
                
            }
        }
    }
    
    private func handleOOBIncomingMessage(_ response: EMFetchMessageResponse!) {
        var title : String!
        
        var positiveButtonLabel : String!
        var negativeButtonLabel : String!
        
        // Defined the dialog UI element base on the message type
        switch response.messageType {
        case .unsupported, .generic, .transactionVerifyRequest:
            //Not handle in this example
            return
        case .transactionSigningRequest:
            positiveButtonLabel = "Approve"
            negativeButtonLabel = "Deny"
            title = "Do you want to accept this message?"
        @unknown default:
            return
        }
        
        let alertController : UIAlertController = UIAlertController.init(title: title, message: nil, preferredStyle: .alert)
        let positiveButton : UIAlertAction = UIAlertAction.init(title: positiveButtonLabel, style: .default) { (_) in
            self.responseToAction(response, value: true)
        }
        let negativeButton : UIAlertAction = UIAlertAction.init(title: negativeButtonLabel, style: .cancel) { (_) in
            self.responseToAction(response, value: false)
        }
        alertController.addAction(positiveButton)
        alertController.addAction(negativeButton)
        self.present(alertController, animated: true, completion: nil)
    }
    
    private func responseToAction(_ response: EMFetchMessageResponse, value: Bool) {
        switch response.messageType {
        case .unsupported, .generic, .transactionVerifyRequest:
            return
        case .transactionSigningRequest:
            // If the message is asking to sign a transaction the user could decide to accept or reject it
            if (value) {
                self.showPinInput { authInput, pin in
                    self.signTransactionMessage(response, value: .accepted, pinAuthInput: authInput, pin: pin)

                }
            } else {
                // Reject the transaction
                self.signTransactionMessage(response, value: .rejected, pinAuthInput: nil, pin: nil)
            }
        @unknown default:
            return
        }
    }
    
    private func signTransactionMessage(_ response: EMFetchMessageResponse,
                                        value: EMTransactionSigningResponseValue,
                                        pinAuthInput: EMProtectorAuthInput?,
                                        pin: String?) {
        self.oOBMessagingLogic.signTransactionMessage(response, value: value, pinAuthInput: pinAuthInput, pin: pin, completion: { (error) in
            if let error = error {
                self.displayMessageDialogError(error: error)
            } else {
                self.displayMessageDialog(result: "This message is \(value == .accepted ? "ACCEPTED" : "REJECTED") successfully\n")
            }
        })
    }
    
}
